ansible/roles/images/tasks/main.yaml

```diff
@@ -7,21 +7,57 @@
 - name: load images
   import_tasks: load.yaml
 
-- name: pull control plane images
-  include_tasks: check-or-pull-images.yaml
-  vars:
-    image_name: "{{ item }}"
-  loop: "{{ control_plane_images }}"
+# NOTE(jkoelker) Grab the build tag (eg. `fips.0` from `v1.21.6+fips.0`)
+- name: detect build
+  shell: kubeadm version -o short | cut -d+ -f2 -s
+  register: kubeadm_build_tag
 
-- name: pull extra images
-  include_tasks: check-or-pull-images.yaml
-  vars:
-    image_name: "{{ item }}"
-  loop: "{{ extra_images }}"
+- name: set kubernetes build tag
+  set_fact:
+    kubernetes_build_tag: "{{ kubeadm_build_tag.stdout | trim }}"
+
+# NOTE(jkoelker) If `kubeadm` has a build tag, append that to the images.
+#                Since kubernets uses a build tag on their `etcd` images (e.g
+#                `-0` from `k8s.gcr.io/etcd:3.4.13-0`), strip it prior to
+#                appending the build tag from `kubeadm`.
+- name: determine kubernetes images
```

Comment:
Can we write out a temp kubeadm config setting the `k8s_image_registry` and `kubernetes_build_tag` instead? This way we'll have kubeadm print us the version thats its going to use when running `init/join`.

Reply:
See https://github.com/mesosphere/konvoy2/blob/main/pkg/capi/generator/testdata/cluster/aws-minimal-etcd-image-ref.yaml#L92-L95

Reply:
We can set the repository that way easy peasy, but we have a chicken and egg with the `etcd` and `coredns` tags, for example the `imageTag: v3.4.13_fips.0` for etcd requires the version, but we are trying to have `kubeadm` inform us the version, how do we know that kubeadm would want (`v3.4.13`)?

We could do a two phase where we first run it without any config and just parse out the etcd and coredns versions and remove the build tag from the etcd version (upstream adds `-0` etc when they rebuild etcd), then write the config, run it again with the registry set and get the results back.

I'll prototype that out, I think that the `sed`s are ugly enough that a bit more gymnastics like that would be justified.

Sidenote it looks like for FIPS we are not building our own coredns with boringssl (I couldn't find it on dockerhub ;( ).